src/v/s3/client.cc

```diff
@@ -376,6 +406,31 @@ ss::future<ResultT> parse_rest_error_response(iobuf&& buf) {
     }
 }
 
+/// Head response doesn't give us an XML encoded error object in
+/// the body. This method uses headers to generate an error object.
+template<class ResultT = void>
```

Comment:
just curious, why do we need to make return type a template argument?

Reply:
> just curious, why do we need to make return type a template argument?

I think it is so that it can be re-used in contexts with different return types like:

```c++
future<A> a() { return parse_head_error_response<A>(...); }
future<B> b() { return parse_head_error_response<B>(...); }
```

is that right @Lazin ?